File: nim/gol.nim
import std/[os,math]
type
  Matrix[W, H: static[int]] =
    array[1..W, array[1..H, int]]

const W = 15
const H = 15
const FPS = 10

proc alive(matrix:Matrix[W,H],x,y:int): int =
    let dirs:Matrix[8,2] = [
            [1,-1],
            [1,0],
            [1,1],
            [-1,-1],
            [-1,0],
            [-1,1],
            [0,1],
            [0,-1],
    ] 
    var count:int = 0
    for dir in dirs:
        var newx =  ( x + dir[1] + W ) mod W
        var newy = ( y + dir[2] + H ) mod H 
        if newx == 0:
            newx = W
        if newy == 0:
            newy = H

        if matrix[newy][newx] == 1:
            count += 1
    return count




proc print(a:Matrix[W,H])  = 
    # clear screen and move cursor to (1,10)[x,y]
    stdout.write "\x1b[2J\x1b[10;1H"
    for row in a:
        stdout.write "\t"
        for col in row:
            if(col == 1):
                stdout.write "⬜"
            else:
                stdout.write "⬛"
        echo("")


proc next_step(matrix:Matrix[W,H]): Matrix[W,H]=
    var buffer:Matrix[W,H];
    for y in low(matrix)..high(matrix):
        for x in  low(matrix[y])..high(matrix[y]):
            case alive(matrix,x,y):
                of 3:
                    buffer[y][x] = 1
                of 2:
                    buffer[y][x] = matrix[y][x]
                else:
                    buffer[y][x] = 0
    return buffer


proc glider(matrix:Matrix[W,H]):Matrix[W,H]=
    var matrix:Matrix[W,H];
    matrix[2][3]=1;
    matrix[3][4]=1;
    matrix[4][4]=1;
    matrix[4][3]=1;
    matrix[4][2]=1;
    return matrix

var matrix:Matrix[W,H];

matrix = glider(matrix)


while 1==1:
    print(matrix)
    matrix = next_step(matrix)
    sleep( int( 1000/FPS))

File: zig/gol.zig
const std = @import("std");
const print = std.debug.print;
const sleep = std.time.sleep;

const ROWS = 50;
const COLS = 50;

const MatrixT = [ROWS][COLS]i32;

pub fn display(matrix:MatrixT) void {
    print("\x1b[1J\x1b[1;1H",.{});
    for (matrix)|row| {
        for (row)|col| {
                if(col==1){
                    print("⬜",.{});
                }else{
                    print("⬛",.{});
                }
        }
        print("\n",.{});
    }
}

pub fn alive(matrix:MatrixT,row:i32,col:i32 ) i32 {
    var count:i32 = 0;
    const dir = [_]i32{-1,0,1};
    for (dir)|dy| {
        for (dir)|dx| {
            if(dy==0 and dx==0) continue;
            const newx:usize = @intCast(@mod(( col + dx + COLS ),  COLS));
            const newy:usize = @intCast(@mod(( row + dy + ROWS ),  ROWS));
            if(matrix[newy][newx]==1){
                count+=1;
            }
        }
    }
    return count;
}


pub fn next_step(matrix:MatrixT) MatrixT {
    var buff:MatrixT=undefined; 
    for(0..matrix.len)|row|{
        for(0..matrix[row].len)|col|{
            const alives = alive(matrix,@intCast(row),@intCast(col));
            buff[row][col] = switch (alives) {
                3 => 1,
                2 => matrix[row][col],
                else => 0,
            };
        }
    }
    return buff;
}


pub fn fill_random() MatrixT {
    var buff:MatrixT = undefined;
    for(0..buff.len)|row|{
        for(0..buff[row].len)|col|{
            const rand = std.crypto.random;
            if(rand.float(f32)>0.5){
                buff[row][col] =1;
            }else{
                buff[row][col] = 0;

            }
        }
    }
    return buff;
}


pub fn main() !void {
    var a :MatrixT = fill_random();
    a[1][2]=1;
    a[2][3]=1;
    a[3][1]=1;
    a[3][2]=1;
    a[3][3]=1;



    //  { 
    //     [_] i32 { 0, 0, 0, 0, 0 },
    //     [_] i32 { 0, 0, 1, 0, 0 },
    //     [_] i32 { 0, 0, 0, 1, 0 },
    //     [_] i32 { 0, 1, 1, 1, 0 },
    //     [_] i32 { 0, 0, 0, 0, 0 },
    // };
    for (0..100000) |_| {
        display(a);
        a = next_step(a);
        sleep(3e+8/ROWS);
    }
}